Sources/AppStoreConnectCLI/Commands/Users/ListUsersCommand.swift

```diff
@@ -32,62 +32,39 @@ public struct ListUsersCommand: CommonParsableCommand {
         help: "Filter the results by the specified username.",
         transform: { $0.lowercased() }
     )
-    var filterUsername: [String?]
+    var filterUsername: [String]
 
     @Option(
         parsing: .upToNextOption,
         help: ArgumentHelp(stringLiteral: "Filter the results by the specified roles (\(UserRole.allCases.map { $0.rawValue.lowercased() }.joined(separator: ", "))).")
     )
-    var filterRole: [UserRole?]
+    var filterRole: [UserRole]
 
     @Option(
         parsing: .upToNextOption,
         help: ArgumentHelp(stringLiteral: "Filter the results by the app(s) visible to each user.\nUsers with access to all apps will always be included."),
         transform: { $0.lowercased() }
     )
-    var filterVisibleApps: [String?]
+    var filterVisibleApps: [String]
 
     @Flag(help: "Include visible apps in results.")
     var includeVisibleApps: Bool
 
     public func run() throws {
-        let api = try makeClient()
-
-        var filters = [ListUsers.Filter]()
-
-        if !filterRole.isEmpty {
-            filters.append(ListUsers.Filter.roles(filterRole.compactMap { $0?.rawValue }))
-        }
-
-        if !filterUsername.isEmpty {
-            filters.append(ListUsers.Filter.username(filterUsername.compactMap { $0 }))
-        }
-
-        if !filterVisibleApps.isEmpty {
-            filters.append(ListUsers.Filter.visibleApps(filterVisibleApps.compactMap { $0 }))
-        }
-
-        let limits = [
-          limitUsers.map(ListUsers.Limit.users),
-          limitVisibleApps.map(ListUsers.Limit.visibleApps)
-        ]
-        .compactMap { $0 }
-
-        let request = APIEndpoint.users(
-            fields: nil,
-            include: includeVisibleApps
-                ? [ListUsers.Include.visibleApps]
-                : nil,
-            limit: limits,
-            sort: [sort].compactMap { $0 },
-            filter: filters,
-            next: nil)
-
-        _ = api.request(request)
-            .map(User.fromAPIResponse)
-            .sink(
-                receiveCompletion: Renderers.CompletionRenderer().render,
-                receiveValue: Renderers.ResultRenderer(format: common.outputFormat).render
-            )
+        let service = try makeService()
+
+        let options = ListUsersOptions(
+            limitVisibleApps: limitVisibleApps,
+            limitUsers: limitUsers,
+            sort: sort,
+            filterUsername: filterUsername,
+            filterRole: filterRole,
+            filterVisibleApps: filterVisibleApps,
+            includeVisibleApps: includeVisibleApps
+        )
+
+        _ = service
```

Comment:
What would be great would be if we could have some sort of async/await or future type construct so we didn't have to write everything as a giant combine chain.

```swift
let users = service.listUsers(with: options).await()
users.renderResult(format: common.outputFormat)
```

This I think would make writing more complex commands simpler.  It would also make the blocking explicit rather than buried in `AppStoreConnectService`.

Reply:
Yeah, I have been thinking about this too. It's a really good idea

Reply:
I'll address this in a different PR